[Logs UI / Metrics UI] New Platform migration server followups

Kerry350 opened the pull request:
~⚠️ ⚠️ ⚠️ **This contains all of the commits of #45299 and shouldn't be reviewed until that is merged**~

## Summary

~This is currently a WIP, but open for visibility.~

This is a followup PR to #45299, and contains changes that were important, but not blockers in getting the other PR merged.

- Adds types for plugin dependencies where possible (0a40478)
- Fully converts Metrics Explorer to using io-ts. This reinstates validation for the Metrics Explorer API which was temporarily disabled (6748dab).
- Exposes our public contract, for other plugins to consume, via the `setup()` method return statement (a918f08) (This isn't used yet, but it just gets us ready for when we move 100% to the NP world / directory).
- Adds a `kibanaVersion` property to the `kibana.json` manifest for the Infra plugin (b9cb4ee).

The changeset here looks a lot bigger than it realistically is. The vast majority of the changes relate to the io-ts conversion for Metrics Explorer. As some of the enums have changed to string unions a lot of files were touched, but for minimal changes.

For testing this PR, please ensure the Metrics portions of the UI still work as expected.

(@elastic/apm-ui have been automatically requested for review as I've removed a comment from an APM file that we left behind in the big migration PR.)

jasonrhodes committed:
Basic cleanup before refactoring for shim work

jasonrhodes committed:
Basic cleanup before refactoring for shim work

jasonrhodes committed:
shim WIP

jasonrhodes committed:
shim WIP

jasonrhodes committed:
Removes the configuration adapter

jasonrhodes committed:
Removes the configuration adapter

jasonrhodes committed:
WIP more stuff

jasonrhodes committed:
WIP more stuff

jasonrhodes committed:
WIP refactoring of shimming work

jasonrhodes committed:
WIP refactoring of shimming work

jasonrhodes committed:
WIP continues

jasonrhodes committed:
WIP continues

jasonrhodes committed:
Logging UI now runs on top of new platform shim

jasonrhodes committed:
Logging UI now runs on top of new platform shim

jasonrhodes committed:
WIP continues

jasonrhodes committed:
WIP continues

jasonrhodes committed:
Removes unused imports and variables

jasonrhodes committed:
Removes unused imports and variables

jasonrhodes committed:
Basic infra NP server shim in place

jasonrhodes committed:
Basic infra NP server shim in place

jasonrhodes committed:
Reimplemented graphql http error handling for infra NP server shim

jasonrhodes committed:
Reimplemented graphql http error handling for infra NP server shim

jasonrhodes committed:
Adds new platform infra plugin to handle NP config for legacy server shim

jasonrhodes committed:
Adds new platform infra plugin to handle NP config for legacy server shim

jasonrhodes committed:
Adds comment about duplicating full config for NP config

Kerry350 committed:
Use New Platform features plugin to registerFeature()

jasonrhodes committed:
Fixing merge conflicts from collaboration

jasonrhodes committed:
Re-arranging and relying on request context as uch as possible

jasonrhodes committed:
Refactors KibanaRequest for RequestHandlerContext

jasonrhodes committed:
fixes types for callWithRequest

jasonrhodes committed:
Moves callWithRequest method override types directly into class to get them working, need to fix this when we understand it better

jasonrhodes committed:
Merge remote-tracking branch 'upstream/master' into logs-ui-server-np-shim

jasonrhodes committed:
Fixes callWithRequest framework types

jasonrhodes committed:
Removes a few NP_TODO comments

Kerry350 committed:
Merge remote-tracking branch 'upstream/master' into logs-ui-server-np-shim

Kerry350 committed:
Fix broken imports

Kerry350 committed:
Ensure GraphQL resolvers are actually passed requestContext and not the raw request, and switch to the savedObjects client via requestContext

Kerry350 committed:
Merge remote-tracking branch 'upstream/master' into logs-ui-server-np-shim

Kerry350 committed:
Remove the legacy traces of the savedObjects plugin

Kerry350 committed:
Fixes TSVB access with NP raw requests and requestContext

Kerry350 committed:
Remove unused getUiSettingsService (moved to requestContext)

Kerry350 committed:
Merge remote-tracking branch 'upstream/master' into logs-ui-server-np-shim

Kerry350 committed:
Migrate to new Spaces plugin

Kerry350 committed:
Fix calculateMetricInterval after merged changes

Kerry350 committed:
Reinstate and migrate the infrastructure metadata route

Kerry350 committed:
Merge remote-tracking branch 'upstream/master' into logs-ui-server-np-shim

Kerry350 committed:
Fix various type check errors

Kerry350 committed:
Amend InfraSources lib unit tests

Mock the savedObjects client differently

Kerry350 committed:
Amend MetricsExplorer API response

Renaming of variable inadvertently broke the response

Kerry350 committed:
Merge remote-tracking branch 'upstream/master' into logs-ui-server-np-shim

Kerry350 committed:
Remove GraphQLI references from feature controls tests

Kerry350 committed:
Remove other GraphiQL references

Kerry350 committed:
Fix security / access issue

Kerry350 committed:
Add a framework level registerRoute method which always adds access tags by default

Kerry350 committed:
*Temp* disable test

Kerry350 committed:
Merge remote-tracking branch 'upstream/master' into logs-ui-server-np-shim

Kerry350 committed:
Merge remote-tracking branch 'upstream/master' into np-migration-followups

Kerry350 committed:
Add real types for plugin dependencies where possible

Kerry350 committed:
Remove comment

It doesn't seem necessary to re-throw the Error

Kerry350 committed:
Adjust some comments that are repeats and will be flagged by TS once parameters aren't needed

Kerry350 committed:
Merge remote-tracking branch 'upstream/master' into np-migration-followups

Kerry350 committed:
Merge remote-tracking branch 'upstream/master' into np-migration-followups

Kerry350 committed:
Merge remote-tracking branch 'upstream/master' into logs-ui-server-np-shim

elasticmachine committed:
Merge branch 'master' into logs-ui-server-np-shim

Kerry350 committed:
Fully convert Metrics Explorer to using io-ts for types / validation

The reinstates commented out validation and removes Joi in favour of io-ts

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/11892/)
* Commit: 6748dabf7eba75bf08ee264d325aff41c45fbbc9

Kerry350 committed:
Merge branch 'logs-ui-server-np-shim' of github.com:jasonrhodes/kibana into logs-ui-server-np-shim

Kerry350 committed:
Merge remote-tracking branch 'upstream/master' into logs-ui-server-np-shim

Kerry350 committed:
Migrate the log rate validation endpoint to the new platform

Fully migrates the [Logs UI] log rate setup index validation #50008 PR to New Platform routing etc

Kerry350 committed:
Merge remote-tracking branch 'jason-fork/logs-ui-server-np-shim' into np-migration-followups

Kerry350 committed:
Merge remote-tracking branch 'upstream/master' into logs-ui-server-np-shim

Kerry350 committed:
Merge remote-tracking branch 'jason-fork/logs-ui-server-np-shim' into np-migration-followups

Kerry350 committed:
Expose public contract to other plugins via setup function

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/12390/)
* Commit: a918f0817730b7db000efc825550190a54c4a018

jasonrhodes committed:
Merge remote-tracking branch 'upstream/master' into logs-ui-server-np-shim

Kerry350 committed:
Merge remote-tracking branch 'upstream/master' into logs-ui-server-np-shim

Kerry350 committed:
Merge branch 'logs-ui-server-np-shim' into np-migration-followups

Kerry350 committed:
Commit types
This should have been in a previous commit

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/12573/)
* Commit: 97c1d143762e06283ace6e14cac753b6cb4ee8d3

Kerry350 committed:
Merge remote-tracking branch 'upstream/master' into logs-ui-server-np-shim

Kerry350 committed:
Amend types

Kerry350 committed:
Merge branch 'logs-ui-server-np-shim' into np-migration-followups

jasonrhodes committed:
Merge branch 'logs-ui-server-np-shim' of github.com:jasonrhodes/kibana into logs-ui-server-np-shim

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/12957/)
* Commit: 34f2c9157ae2a9448d09ed883f2f1f9f384166ea

Kerry350 committed:
Fix type

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/13158/)
* Commit: c61c821fca7fbc8814933b4133c61a3db2422d25

jasonrhodes committed:
Example of how to expose APM get indices method in NP

Kerry350 committed:
Fix calls to TSVB bug caused by object mutation
This is a temp fix as the TSVB NP migration will supercede this

Kerry350 committed:
Merge remote-tracking branch 'upstream/master' into logs-ui-server-np-shim

Kerry350 committed:
Merge branch 'logs-ui-server-np-shim' into np-migration-followups

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/13233/)
* Commit: 023c8dc85cfee1bd9f3b65ebb874c0168162cbbe

jasonrhodes committed:
Converts getApmIndices function to accept saved object client, implements usage in infra

jasonrhodes committed:
Merge branch 'logs-ui-server-np-shim' of github.com:jasonrhodes/kibana into logs-ui-server-np-shim

Kerry350 committed:
Merge remote-tracking branch 'upstream/master' into logs-ui-server-np-shim

Kerry350 committed:
Merge branch 'logs-ui-server-np-shim' of github.com:jasonrhodes/kibana into logs-ui-server-np-shim

Kerry350 committed:
Fix APM setup_request tests

jasonrhodes committed:
Fixes some unused references for linting

Kerry350 committed:
Merge remote-tracking branch 'upstream/master' into logs-ui-server-np-shim

Kerry350 committed:
Migrate all work from #50730 to NP

Kerry350 committed:
Remove duplicate declaration files for rison_node and add a single source of truth at x-pack/typings/rison_node.d.ts for x-pack uses

Kerry350 committed:
Merge remote-tracking branch 'upstream/master' into logs-ui-server-np-shim

Kerry350 committed:
Merge branch 'logs-ui-server-np-shim' into np-migration-followups

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/13882/)
* Commit: 8f2488a7a44f4173f61b661663a041f82092f251

Kerry350 committed:
Merge remote-tracking branch 'upstream/master' into np-migration-followups

Kerry350 committed:
Sort typings file (conflict resolution mistake)

Kerry350 committed:
Merge remote-tracking branch 'upstream/master' into np-migration-followups

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14169/)
* Commit: 2516905390acc53f1e21cfbef5aff0ccb12ccf74

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14169/","number":"14169","commit":"2516905390acc53f1e21cfbef5aff0ccb12ccf74"}]}
PIPELINE-->

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14185/)
* Commit: 4bd6fc5b60a4410e4c91a47f00786724cbe7a5d9

### History
* :broken_heart: [Build #14169](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14169/) failed 2516905390acc53f1e21cfbef5aff0ccb12ccf74

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14169/","number":"14169","commit":"2516905390acc53f1e21cfbef5aff0ccb12ccf74"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14185/","number":"14185","commit":"4bd6fc5b60a4410e4c91a47f00786724cbe7a5d9"}]}
PIPELINE-->

Kerry350 committed:
Merge remote-tracking branch 'upstream/master' into np-migration-followups

Kerry350 committed:
Add kibanaVersion property to kibana.json manifest

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14685/)
* Commit: b9cb4ee566d4b1a66bde964a187d61f9f3ad77ac

### History
* :green_heart: [Build #14185](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14185/) succeeded 4bd6fc5b60a4410e4c91a47f00786724cbe7a5d9
* :broken_heart: [Build #14169](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14169/) failed 2516905390acc53f1e21cfbef5aff0ccb12ccf74

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14169/","number":"14169","commit":"2516905390acc53f1e21cfbef5aff0ccb12ccf74"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14185/","number":"14185","commit":"4bd6fc5b60a4410e4c91a47f00786724cbe7a5d9"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14685/","number":"14685","commit":"b9cb4ee566d4b1a66bde964a187d61f9f3ad77ac"}]}
PIPELINE-->

simianhacker commented on x-pack/legacy/plugins/infra/common/http_api/metrics_explorer/index.ts:
```diff
@@ -0,0 +1,88 @@
+
+import * as rt from 'io-ts';
+
+export const metricsExplorerAggregationRT = rt.keyof({
```
I wonder if we should create an array for each of these aggregations
```
const METRIC_EXPLORER_AGGREGATIONS = ['avg', 'max', 'min', 'cardinality', 'rate', 'count'];
```
then define `metricsExplorerAggregagtionRT` as
```
export const metricsExplorerAggregationRT = rt.keyof(
  METRIC_EXPLORER_AGGREGATIONS.reduce(
  (acc, agg) => ({ ...acc, [agg]: null })
  , {})
);
```
Then in x-pack/legacy/plugins/infra/public/components/metrics_explorer/aggregation.tsx we can just include `METRICS_EXPLORER_AGGREGATIONS` and use it on [line 68](https://github.com/elastic/kibana/pull/51615/files#diff-e5ba648c4dc9fdcb312a0ba814e0bb0dR68)

simianhacker replied:
That reduce might not work, I think io-ts needs that to be a literal object instead of a dynamic one.

simianhacker commented on x-pack/legacy/plugins/infra/public/components/metrics_explorer/aggregation.tsx:
```diff
@@ -18,42 +18,41 @@ interface Props {
-      const aggregation =
-        (isMetricsExplorerAggregation(e.target.value) && e.target.value) ||
-        MetricsExplorerAggregation.avg;
+      const aggregation = (isMetricsExplorerAggregation(e.target.value) && e.target.value) || 'avg';
```
See note above.

simianhacker commented on x-pack/legacy/plugins/infra/public/components/metrics_explorer/aggregation.tsx:
```diff
@@ -66,7 +65,7 @@ export const MetricsExplorerAggregationPicker = ({ options, onChange }: Props) =
       fullWidth
       value={options.aggregation}
-      options={Object.keys(MetricsExplorerAggregation).map(k => ({
+      options={['avg', 'max', 'min', 'cardinality', 'rate', 'count'].map(k => ({
```
See note above.

simianhacker reviewed (changes requested):
Everything looks good. The only thing that stuck out to me is around the aggregation type. It seems like we are defining that list several times so I added some suggestions to try and reduce that some. If you want to just move forward with what you have, just submit for review again and I will approve.

jasonrhodes reviewed (approved):
I like Chris's dedupe suggestions but otherwise this LGTM

jasonrhodes commented on x-pack/legacy/plugins/apm/server/lib/settings/apm_indices/get_apm_indices.ts:
```diff
@@ -54,10 +54,6 @@ export function getApmIndicesConfig(config: APMConfig): ApmIndicesConfig {
   };
 }
 
-// export async function getApmIndices(context: APMRequestHandlerContext) {
```
lol sorry, I think this was my fault -- thanks for the clean up :) <3

Kerry350 committed:
Merge remote-tracking branch 'upstream/master' into np-migration-followups

Kerry350 committed:
Remove duplication around metrics explorer aggregation types

Kerry350 commented:
@simianhacker Thanks for the review and suggestions. I was going to do these in a followup, but they didn't take very long so I've added them in 2139171. If you're happy and can approve I'll merge this 👍

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/15815/)
* Commit: d96a4f67df7b6af277d182f21c8ab4843cdf27ad

### History
* :green_heart: [Build #14685](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14685/) succeeded b9cb4ee566d4b1a66bde964a187d61f9f3ad77ac
* :green_heart: [Build #14185](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14185/) succeeded 4bd6fc5b60a4410e4c91a47f00786724cbe7a5d9
* :broken_heart: [Build #14169](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14169/) failed 2516905390acc53f1e21cfbef5aff0ccb12ccf74

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14169/","number":"14169","commit":"2516905390acc53f1e21cfbef5aff0ccb12ccf74"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14185/","number":"14185","commit":"4bd6fc5b60a4410e4c91a47f00786724cbe7a5d9"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14685/","number":"14685","commit":"b9cb4ee566d4b1a66bde964a187d61f9f3ad77ac"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/15815/","number":"15815","commit":"d96a4f67df7b6af277d182f21c8ab4843cdf27ad"}]}
PIPELINE-->

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/15834/)
* Commit: 21391719b8dc9c62e670e74fb706750602b9c364

### History
* :green_heart: [Build #15815](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/15815/) succeeded d96a4f67df7b6af277d182f21c8ab4843cdf27ad
* :green_heart: [Build #14685](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14685/) succeeded b9cb4ee566d4b1a66bde964a187d61f9f3ad77ac
* :green_heart: [Build #14185](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14185/) succeeded 4bd6fc5b60a4410e4c91a47f00786724cbe7a5d9
* :broken_heart: [Build #14169](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14169/) failed 2516905390acc53f1e21cfbef5aff0ccb12ccf74

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14169/","number":"14169","commit":"2516905390acc53f1e21cfbef5aff0ccb12ccf74"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14185/","number":"14185","commit":"4bd6fc5b60a4410e4c91a47f00786724cbe7a5d9"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14685/","number":"14685","commit":"b9cb4ee566d4b1a66bde964a187d61f9f3ad77ac"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/15815/","number":"15815","commit":"d96a4f67df7b6af277d182f21c8ab4843cdf27ad"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/15834/","number":"15834","commit":"21391719b8dc9c62e670e74fb706750602b9c364"}]}
PIPELINE-->

simianhacker reviewed (approved):
LGTM
